review solid and clean code

Shafagh-Sepehr opened the pull request.

Shafagh-Sepehr committed:
add

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+    private static void Main()
+    {
+        
+        var invertedIndex = new InvertedIndexDictionary("/home/shafagh/Desktop/EnglishData", ["will",]);
```
آدرس رو هارد کد بذاریم؟‌ راه بهتر چی هست؟ چرا؟

Shafagh-Sepehr replied:
یه فایل appsettings.json درست کردم گذاشتم اونجا.
تا هم اگه خواست عوض بشه جاش معلوم باشه هم فرضا اگه برنامه بزرگتر شد یه جا برای اینطور مقادیر داشته باشیم

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+        var invertedIndex = new InvertedIndexDictionary("/home/shafagh/Desktop/EnglishData", ["will",]);
+
+
+        Console.Write("Search: ");
```
کلا برا خروجی ها چطوره یه کلاس داشته باشیم؟

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+
+
+        Console.Write("Search: ");
+        string? query = Console.ReadLine();
```
و همچنین ورودی گرفتن‌ها؟

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+        Console.Write("Search: ");
+        string? query = Console.ReadLine();
+
+        if (string.IsNullOrWhiteSpace(query)) return;
```
خطا مشخص نیست 
write new exception

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+        IEnumerable<string> result = invertedIndex.Search(query);
+
+        Console.WriteLine("Result:");
+        foreach (string doc in result)
```
{
}

navidnasyry replied:
use Implicitly type. why?

navidnasyry commented on FullTextSearch/Program.cs:
```diff
@@ -0,0 +1,22 @@
+
+        Console.WriteLine("Result:");
+        foreach (string doc in result)
+            Console.WriteLine(doc);
```
use class and interface

navidnasyry commented on FullTextSearch/InvertedIndexDictionary.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        var stringToWordsProcessor = new StringToWordsProcessor(banned);
+        var englishPorter2Stemmer = new EnglishPorter2Stemmer();
+        invertedIndexDictionaryBuilder ??= new InvertedIndexDictionaryBuilder(stringToWordsProcessor, englishPorter2Stemmer);
```
این کلاس به چه کلاس‌های دیگه وابستس؟ این خوبه؟

navidnasyry commented on FullTextSearch/InvertedIndexDictionaryBuilder.cs:
```diff
@@ -0,0 +1,41 @@
+
+    public Dictionary<string, List<string>> Build(string filepath)
+    {
+        string[] files = Directory.GetFiles(filepath);
```
var

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+        _invertedIndex = invertedIndex;
+
+
+        if (OrWords.Count > 0 && AndWords.Count > 0)
+            result = GetDocumentsForAndQueries().Intersect(GetDocumentsForOrQueries())
+                .Except(GetDocumentsForNotQueries());
+        else if (AndWords.Count > 0)
+            result = GetDocumentsForAndQueries().Except(GetDocumentsForNotQueries());
+        else if (OrWords.Count > 0)
+            result = GetDocumentsForOrQueries().Except(GetDocumentsForNotQueries());
+        
```
use blocks

navidnasyry replied:
هر کدوم از این کارا میتونن یه سرویس جدا بشن

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+
+namespace FullTextSearch;
+
+internal class QuerySearcher(IPorter2Stemmer? stemmer = null) : IQuerySearcher
```
این کلاس خیلی طولانیه که داره همه کاری میکنه:)

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+        return result;
+    }
+    
+    private List<string> AndWords
+    {
+        get
+        {
+            return _queryWords
+                .Where(x => x[0] != '+' && x[0] != '-')
+                .Select(x => _stemmer.Stem(x).Value)
+                .ToList();
+        }
+    }
+
```
یه سرویس جدا که اینا رو ارائه بده

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+
+    private List<string> OrWords
+    {
+        get
+        {
+            return _queryWords
+                .Where(x => x[0] == '+')
+                .Select(x => x.Substring(1, x.Length - 1))
+                .Select(x => _stemmer.Stem(x).Value)
+                .ToList();
+        }
+    }
```
ایضا

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+                .ToList();
+        }
+    }
+
+    private List<string> NotWords
+    {
+        get
+        {
+            return _queryWords
+                .Where(x => x[0] == '-')
+                .Select(x => x.Substring(1, x.Length - 1))
+                .Select(x => _stemmer.Stem(x).Value)
+                .ToList();
+        }
+    }
```
..

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+    }
+
+
+    private HashSet<string> GetDocumentsForAndQueries()
+    {
+        List<List<string>> andDocsList = _invertedIndex
+            .Where(x => AndWords.Contains(x.Key))
+            .Select(x => x.Value).ToList();
+
+        if (andDocsList.Count < 1) return [];
+            
+        var result = new HashSet<string>(andDocsList[0]);
+        for (var i = 1; i < andDocsList.Count; i++)
+            result.IntersectWith(andDocsList[i]);
+
+        return result;
+    }
```
another class

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+        return result;
+    }
+
+    private HashSet<string> GetDocumentsForOrQueries()
+    {
+        return _invertedIndex
+            .Where(x => OrWords.Contains(x.Key))
+            .Select(x => x.Value).SelectMany(x => x)
+            .ToHashSet();
+    }
```
...

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+            .ToHashSet();
+    }
+
+    private HashSet<string> GetDocumentsForNotQueries()
+    {
+        return _invertedIndex
+            .Where(x => NotWords.Contains(x.Key))
+            .Select(x => x.Value).SelectMany(x => x)
+            .ToHashSet();
+    }
+}
```
...

navidnasyry commented on FullTextSearch/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,65 @@
+
+internal partial class StringToWordsProcessor : IStringToWordsProcessor
+{
+    private readonly List<string> _banned =
+    [
+        "i", "me", "my", "myself", "we", "our", "ours", "ourselves", "you", "your", "yours", "yourself",
+        "yourselves", "he", "him", "his", "himself", "she", "her", "hers", "herself", "it", "its", "itself", "they",
+        "them", "their", "theirs", "themselves", "what", "which", "who", "whom", "this", "that", "these", "those",
+        "am", "is", "are", "was", "were", "be", "been", "being", "have", "has", "had", "having", "do", "does",
+        "did", "doing", "a", "an", "the", "and", "but", "if", "or", "because", "as", "until", "while", "of", "at",
+        "by", "for", "with", "about", "against", "between", "into", "through", "during", "before", "after", "above",
+        "below", "to", "from", "up", "down", "in", "out", "on", "off", "over", "under", "again", "further", "then",
+        "once", "here", "there", "when", "where", "why", "how", "all", "any", "both", "each", "few", "more", "most",
+        "other", "some", "such", "no", "nor", "not", "only", "own", "same", "so", "than", "too", "very", "s", "t",
+        "can", "will", "just", "don", "should", "now",
+    ];
```
راه بهتر بجای هارد کد کردن

navidnasyry commented on FullTextSearch/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,65 @@
+    public IEnumerable<string> TrimSplitAndStemString(string source, IPorter2Stemmer stemmer)
+    {
+        
+        return source
+            .Trim()
+            .Split()
+            .Where(x => !UrlRegex().IsMatch(x) && !EmailRegex().IsMatch(x) && !PhoneNumberRegex().IsMatch(x))
+            .Select(CleanseString).SelectMany(x => x) // flatten the string arrays
+            .Select(x => stemmer.Stem(x).Value.ToLower())
+            .Where(x => !string.IsNullOrWhiteSpace(x) && x.Length >= 3 && !_banned.Contains(x))
+            .Distinct();
```
tooooo long:(

navidnasyry replied:
extract method and variable

navidnasyry commented on FullTextSearch/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,65 @@
+        "can", "will", "just", "don", "should", "now",
+    ];
+    
+    [GeneratedRegex(@"(https:\/\/www\.|http:\/\/www\.|https:\/\/|http:\/\/)?[a-zA-Z]{2,}(\.[a-zA-Z]{2,})(\.[a-zA-Z]{2,})?\/[a-zA-Z0-9]{2,}|((https:\/\/www\.|http:\/\/www\.|https:\/\/|http:\/\/)?[a-zA-Z]{2,}(\.[a-zA-Z]{2,})(\.[a-zA-Z]{2,})?)|(https:\/\/www\.|http:\/\/www\.|https:\/\/|http:\/\/)?[a-zA-Z0-9]{2,}\.[a-zA-Z0-9]{2,}\.[a-zA-Z0-9]{2,}(\.[a-zA-Z0-9]{2,})?")]
+    private partial Regex UrlRegex();
+    [GeneratedRegex("(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a ...
+    private partial Regex EmailRegex();
+    [GeneratedRegex(@"^[\+]?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$")]
```
خوانا نیست...چرا نیازه؟

navidnasyry commented on FullTextSearch/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,65 @@
+    
+    private string[] CleanseString(string value)
+    {
+        value = value.Trim(":.,<>?/\\\"\t'`~!@#$%^&*()_+=-*;\t |".ToCharArray()); //trim noise around probably important words
```
کد اونقدر باید خوانا باشه که نیاز به کامنت نداشته باشه:)
حذف کامنت و خوانا تر کردن با متغیر و متد و کلاس

navidnasyry commented on FullTextSearch/StringToWordsProcessor.cs:
```diff
@@ -0,0 +1,65 @@
+    {
+        value = value.Trim(":.,<>?/\\\"\t'`~!@#$%^&*()_+=-*;\t |".ToCharArray()); //trim noise around probably important words
+        
+        if (!IsNumberAndMoreThan2Digits(value))
+            value = value.Trim("012345689".ToCharArray()); //trim digits of a string if the string isn't at least a two-digit number
+        
```
"""

navidnasyry commented on FullTextSearch/InvertedIndexDictionaryBuilder.cs:
```diff
@@ -0,0 +1,41 @@
+
+            IEnumerable<string> words = _stringToWordsProcessor.TrimSplitAndStemString(content, _stemmer);
+
+            foreach (string word in words)
```
{}

navidnasyry replied:
extract method

navidnasyry commented on FullTextSearch/QuerySearcher.cs:
```diff
@@ -0,0 +1,101 @@
+    private readonly IPorter2Stemmer                  _stemmer       = stemmer ?? new EnglishPorter2Stemmer();
+
+
+    public IEnumerable<string> Search(string query, Dictionary<string, List<string>> invertedIndex)
```
S in Solid

navidnasyry reviewed (changes requested):
Read more about Design Patterns :

Singleton
Factory
Strategy
and ...

navidnasyry reviewed (commented):
و راستی...
پوشه بندی هم میتونه بهتر باشه بر اساس کار‌هایی که انجام میشه
میتونی 
Clean Architecture و MVC
رو هم بخونی

Shafagh-Sepehr committed:
move hard coded path to a newly created appsettings.json file.

Shafagh-Sepehr committed:
create new exception for null or whitespace input.

Shafagh-Sepehr committed:
create UserOutput class for output related actions.

Shafagh-Sepehr committed:
create UserInput class for input related actions.

Shafagh-Sepehr committed:
refactor UserInput and UserOutput classes name and methods and also extract interface.

Shafagh-Sepehr committed:
replace appsettings.json with a public static class for better accessibility across the project.

Shafagh-Sepehr committed:
move hardcoded banned words to the AppSettings class.

Shafagh-Sepehr committed:
refactor StringToWordsProcessor's TrimSplitAndStemString function:
extract two methods from it
get the stemmer via constructor instead of this method

Shafagh-Sepehr committed:
extract methods from TrimSplitAndStemString method

Shafagh-Sepehr committed:
change folder structure

Shafagh-Sepehr committed:
move InvertedIndex folder to a new Application folder.

Shafagh-Sepehr committed:
extract words processing from QuerySearcher.cs

Shafagh-Sepehr committed:
change QueryProcessor's name to WordProcessor

Shafagh-Sepehr committed:
refactored Query searcher.
It now inputs the inverted index through constructor.

Shafagh-Sepehr committed:
extract class from QuerySearcher.cs
GetDocumentForXQuery methods are now each a class and there's a master DocumentReader class that encapsulates them.

Shafagh-Sepehr committed:
extract methods from QuerySearcher's Search method

Shafagh-Sepehr committed:
improve inject IDocumentReader to QuerySearcher class for future test mocks.

Shafagh-Sepehr committed:
rename WordProcessor folder to WordProcessors because a class with the same name exists in it.

Shafagh-Sepehr committed:
fix the bug in QuerySearcher where Words are passed to document readers before initialization.

Shafagh-Sepehr committed:
extract methods from Search method of QuerySearcher class.

Shafagh-Sepehr committed:
rename Search methods to better ones.

Shafagh-Sepehr committed:
extract method from InvertedIndexDictionaryBuilder's FillInvertedIndexFromFile method.

Shafagh-Sepehr committed:
create builder class for InvertedIndexDictionary to decouple it from dependencies and also making instances easier.

Shafagh-Sepehr committed:
move regex patterns to AppSettings class

Shafagh-Sepehr committed:
Add Dependency Injection to the project for all classes.
create a Construct method to pass the needed parameters.
remove unused InvertedIndexBuilder.cs.
info: tests are down.

Shafagh-Sepehr committed:
Capitalize first letter of the Message in NullInputException.

Shafagh-Sepehr committed:
Add Assertions for classes that their Construct method should be called.

Shafagh-Sepehr committed:
extract Search methods to classes in QuerySearcher

Shafagh-Sepehr committed:
delete unused using statements.

Shafagh-Sepehr committed:
encapsulate class parameters

Shafagh-Sepehr committed:
remove interface postfix from interface namespaces.

Shafagh-Sepehr committed:
create InvertedIndexFactory for ease of use by both the user and developer.
extract Services to its own class.

Shafagh-Sepehr committed:
hide classes from user by changing them from public to internal.
